ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,154 @@
+func (t *tracer) reportAbandonedSpans(interval time.Duration) {
+	tick := time.NewTicker(tickerInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			logAbandonedSpans(t.abandonedSpans, true, interval)
+		case s := <-t.cIn:
+			bNode := t.abandonedSpans.Front()
+			if bNode == nil {
+				b := list.New()
+				b.PushBack(s)
+				t.abandonedSpans.PushBack(b)
+				break
+			}
+			for bNode != nil {
+				bucket, bOk := bNode.Value.(*list.List)
+				if !bOk {
+					bNode = bNode.Next()
+					continue
+				}
+				sNode := bucket.Front()
+				if sNode == nil {
+					bNode = bNode.Next()
+					continue
+				}
+				sp, sOk := sNode.Value.(*span)
+				if sOk && sp != nil && s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.PushBack(s)
+					break
+				}
+				bNode = bNode.Next()
+			}
+			if bNode != nil {
+				break
+			}
+			b := list.New()
+			b.PushBack(s)
+			t.abandonedSpans.PushBack(b)
+		case s := <-t.cOut:
+			for node := t.abandonedSpans.Front(); node != nil; node = node.Next() {
+				bucket, ok := node.Value.(*list.List)
+				if !ok || bucket.Front() == nil {
+					continue
+				}
+				spNode := bucket.Front()
+				sp, ok := spNode.Value.(*span)
+				if !ok {
+					continue
+				}
+				if s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.Remove(spNode)
+					if bucket.Front() == nil {
+						t.abandonedSpans.Remove(node)
+					}
+					break
+				}
+			}
+		case <-t.stop:
+			logAbandonedSpans(t.abandonedSpans, false, interval)
+			return
+		}
+	}
+}
+
+// logAbandonedSpans returns a string containing potentially abandoned spans. If `filter` is true,
+// it will only return spans that are older than the provided time `interval`. If false,
+// it will return all unfinished spans.
+func logAbandonedSpans(l *list.List, filter bool, interval time.Duration) {
+	var sb strings.Builder
+	nowTime := now()
+	spanCount := 0
+	truncated := false
+
+	for bucketNode := l.Front(); bucketNode != nil; bucketNode = bucketNode.Next() {
+		bucket, ok := bucketNode.Value.(*list.List)
+		if !ok || bucket == nil {
+			continue
+		}
+
+		// since spans are bucketed by time, finding a bucket that is newer
+		// than the allowed time interval means that all spans in this bucket
+		// and future buckets will be younger than `interval`, and thus aren't
+		// worth checking.
+		if filter {
+			spanNode := bucket.Front()
+			if spanNode == nil {
+				continue
+			}
+			sp, ok := spanNode.Value.(*span)
+			if !ok || sp == nil {
+				continue
+			}
+			if nowTime-sp.Start < interval.Nanoseconds() {
+				continue
+			}
+		}
+		for spanNode := bucket.Front(); spanNode != nil; spanNode = spanNode.Next() {
+			sp, ok := spanNode.Value.(*span)
+			if !ok || sp == nil {
+				continue
+			}
+
+			// despite quitting early, spans within the same bucket can still fall on either side
+			// of the timeout. We should still check if the span is too old or not.
+			if filter && nowTime-sp.Start < interval.Nanoseconds() {
+				break
+			}
```

Comment:
Can we just do this check for the bucket where some will be listed and some won't be, and not do it for the ones where we know for sure that everything in the bucket can be printed?

e.g. you could have a function that prints a single span, and another function that prints an entire bucket (where it loops through all spans in the bucket and calls the first helper function). Then just the first bucket can have this item-by-item check, and the remaining buckets can be looped over, and the entire contents printed.

Reply:
Good idea! I think it might be the other way around, though (i.e. the first few buckets should be printed in entirety since they are older, then a later bucket will only be partially printed).